File: AlgorithmsPython/SortingAlgorithms/InsertionSort.py
'''
Created on May 3, 2020

@author: Sarthak Jain
'''

'''
Method that sorts an array using Insertion Sort
'''
def insertionSort(arr):
# Runs till the end of the array's length
    for i in range(1, len(arr)):
        # Sets the key as the value at the index i of an array
        key = arr[i]
        
        # Sets j to one less than i to account for index 0
        j = i - 1
        
        # Loop that runs while j is greater than equal to 0 and the key is less than the value at index j in the array
        while j >= 0 and key < arr[j]:
            # Sets the value at index j+1 to the value of index j 
            arr[j + 1] = arr[j]
        
            # Decrements j by 1
            j = j - 1
        
        # Sets the value at index j + 1 to the key
        arr[j + 1] = key

'''
Runs the main method where it sorts the array and prints out the sorted array
'''
if __name__ == "__main__": 
    print("Sorting an Array using InsertionSort: ")
    arr = [24, 3476984, 11, 22, 69, 0, 88, 34, 331, 1, 23, 732, -1, 4325, 32435432532]
    print("Unsorted Array: " + str(arr));
    insertionSort(arr)
    print("Sorted Array: " + str(arr))
        

File: AlgorithmsJava/src/sortingAlgorithms/InsertionSort.java
package sortingAlgorithms;

import java.util.Arrays;
import java.util.Comparator;

/**
 * Performs an Insertion Sort on a Generic Array
 * 
 * @author Sarthak Jain
 */
public class InsertionSort {
	/**
	 * Runs the main method
	 */
	public static void main(String[] args) {
		Integer[] arr = new Integer[] { 24, 3476984, 11, 22, 69, 0, 88, 34, 331, 1, 23, 732, -1, 4325, 324352532 };
		System.out.println("Unsorted Array before Insertion Sort: " + Arrays.asList(arr));
		insertionSort(arr);
		System.out.println("Sorted Array after Insertion Sort: " + Arrays.asList(arr));
	}

	/**
	 * A generic method that performs an insertion sort on a generic array
	 */
	public static <T extends Comparable<? super T>> void insertionSort(T[] arr) {
		// Creates a for loop that iterates through the array length
		for (int i = 1; i < arr.length; i++) {

			// Creates a temporary value at index at array
			T val = arr[i];

			// Creates another for loop that compares the values of the array at index j and
			// at the temporary value, and compares if the array at index j is greater than
			// the value, then it enters this loop
			for (int j = i - 1; j >= 0 && arr[j].compareTo(val) > 0; j--) {
				// Switches the values in the array
				arr[j + 1] = arr[j];
				arr[j] = val;
			}
		}
	}

}
